### PR TITLE
Java: Model Spring StringUtils

### PR DESCRIPTION
**Pull request by sauyon**
Hopefully these aren't too awful :sweat_smile:

### CODE REVIEW

File: java/ql/src/semmle/code/java/frameworks/spring/SpringStringUtils.qll [hunk 1]
```diff
@@ -0,0 +1,58 @@
+import java
+private import semmle.code.java.dataflow.ExternalFlow
+
+private class SpringStringUtilsModel extends SummaryModelCsv {
+  override predicate row(string row) {
+    row = [
+        "org.springframework.util;StringUtils;false;addStringToArray;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;StringUtils;false;addStringToArray;;;Argument[1];ReturnValue;taint",
+        "org.springframework.util;StringUtils;false;addStringToArray;;;Argument[1];Argument[0];taint",
```
Comment (smowton) [inline java/ql/src/semmle/code/java/frameworks/spring/SpringStringUtils.qll:12]: `addStringToArray` copies the input, so this one doesn't hold

File: java/ql/src/semmle/code/java/frameworks/spring/SpringStringUtils.qll [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        "org.springframework.util;StringUtils;false;applyRelativePath;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;StringUtils;false;applyRelativePath;;;Argument[1];ReturnValue;taint",
```
Comment (smowton) [inline java/ql/src/semmle/code/java/frameworks/spring/SpringStringUtils.qll:14]:
```suggestion
        "org.springframework.util;StringUtils;false;applyRelativePath;;;Argument[0..1];ReturnValue;taint",
```

and similarly any multiple args flowing to the same target

File: java/ql/src/semmle/code/java/frameworks/spring/SpringStringUtils.qll [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        "org.springframework.util;StringUtils;false;arrayToCommaDelimitedString;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;StringUtils;false;arrayToDelimitedString;;;Argument[0];ReturnValue;taint", // for review: this is "org.springframework.util;StringUtils;false;deliminate array with string argument[1]"; should I consider the delimiter to taint the result as well?
```
Comment (smowton) [inline java/ql/src/semmle/code/java/frameworks/spring/SpringStringUtils.qll:16]: It goes in there, so yes (imagine if some careless person allowed a remote user to say whether the sep should be `,` or `;` or `;"; DROP TABLE universe`). Same for other functions inserting a delimiter unless the delimiter is a char

File: java/ql/src/semmle/code/java/frameworks/spring/SpringStringUtils.qll [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        "org.springframework.util;StringUtils;false;concatenateStringArrays;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;StringUtils;false;concatenateStringArrays;;;Argument[1];ReturnValue;taint",
+        "org.springframework.util;StringUtils;false;delete;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;StringUtils;false;deleteAny;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;StringUtils;false;delimitedListToStringArray;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;StringUtils;false;getFilename;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;StringUtils;false;getFilenameExtension;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;StringUtils;false;mergeStringArrays;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;StringUtils;false;mergeStringArrays;;;Argument[1];ReturnValue;taint",
+        "org.springframework.util;StringUtils;false;parseLocale;;;Argument[0];ReturnValue;taint", // for review: do we care about taint through a `Locale`?
```
Comment (smowton) [inline java/ql/src/semmle/code/java/frameworks/spring/SpringStringUtils.qll:34]: Not entirely sure but I think probably not considering it's likely validated on conversion

File: java/ql/test/stubs/springframework-5.2.3/org/springframework/util/StringUtils.java [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+import java.util.Properties;
+import java.util.Set;
+import java.util.StringJoiner;
+import java.util.StringTokenizer;
+import java.util.TimeZone;
+import org.springframework.lang.Nullable;
+
+
+public abstract class StringUtils {
+	public static boolean isEmpty(@Nullable Object str) {
```
Comment (smowton) [inline java/ql/test/stubs/springframework-5.2.3/org/springframework/util/StringUtils.java:40]: Formatting has gone weird here

File: java/ql/src/semmle/code/java/frameworks/spring/SpringStringUtils.qll (hunk 2 above)
Comment (sauyon) [reply java/ql/src/semmle/code/java/frameworks/spring/SpringStringUtils.qll:14]: Is there notation for 0 and 3?

File: java/ql/src/semmle/code/java/frameworks/spring/SpringStringUtils.qll (hunk 1 above)
Comment (sauyon) [reply java/ql/src/semmle/code/java/frameworks/spring/SpringStringUtils.qll:12]: Well that just seems like poor practice :joy:

File: java/ql/src/semmle/code/java/frameworks/spring/SpringStringUtils.qll [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        "org.springframework.util;StringUtils;false;arrayToCommaDelimitedString;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;StringUtils;false;arrayToDelimitedString;;;Argument[0];ReturnValue;taint", // for review: this is "org.springframework.util;StringUtils;false;deliminate array with string argument[1]"; should I consider the delimiter to taint the result as well?
+        "org.springframework.util;StringUtils;false;capitalize;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;StringUtils;false;cleanPath;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;StringUtils;false;collectionToCommaDelimitedString;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;StringUtils;false;collectionToDelimitedString;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;StringUtils;false;collectionToDelimitedString;(java.util.Collection,java.lang.String,java.lang.String,java.lang.String);;Argument[2];ReturnValue;taint", // for review: collectionToDelimitedString has either 2 or 4 parameters; can I just omit the signature here?
```
Comment (sauyon) [inline java/ql/src/semmle/code/java/frameworks/spring/SpringStringUtils.qll:21]: What about this?

File: java/ql/src/semmle/code/java/frameworks/spring/SpringStringUtils.qll (hunk 2 above)
Comment (smowton) [reply java/ql/src/semmle/code/java/frameworks/spring/SpringStringUtils.qll:14]: Fraid not

File: java/ql/src/semmle/code/java/frameworks/spring/SpringStringUtils.qll (hunk 6 above)
Comment (smowton) [reply java/ql/src/semmle/code/java/frameworks/spring/SpringStringUtils.qll:21]: It'll work; @aschackmull do you have a preference on style here?

File: java/ql/src/semmle/code/java/frameworks/spring/SpringStringUtils.qll (hunk 4 above)
Comment (sauyon) [reply java/ql/src/semmle/code/java/frameworks/spring/SpringStringUtils.qll:34]: Removed.

File: java/ql/src/semmle/code/java/frameworks/spring/SpringStringUtils.qll [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        "org.springframework.util;StringUtils;false;quoteIfString;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;StringUtils;false;removeDuplicateStrings;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;StringUtils;false;replace;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;StringUtils;false;replace;;;Argument[2];ReturnValue;taint",
+        "org.springframework.util;StringUtils;false;sortStringArray;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;StringUtils;false;split;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;StringUtils;false;splitArrayElementsIntoProperties;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;StringUtils;false;stripFilenameExtension;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;StringUtils;false;tokenizeToStringArray;;;Argument[0];ReturnValue;taint",
+        "org.springframework.util;StringUtils;false;toLanguageTag;;;Argument[0];ReturnValue;taint",
```
Comment (Marcono1234) [inline java/ql/src/semmle/code/java/frameworks/spring/SpringStringUtils.qll:38]: If `toLanguageTag` is considered, should [`parseLocale`](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/util/StringUtils.html#parseLocale-java.lang.String-) and [`parseLocaleString`](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/util/StringUtils.html#parseLocaleString-java.lang.String-) then be considered as well?

File: java/ql/test/library-tests/frameworks/spring/Test.java [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+    sink(StringUtils.getFilenameExtension(taint())); // $hasTaintFlow
+
+    sink(StringUtils.mergeStringArrays(taintArray(), null)); // $hasTaintFlow
+    sink(StringUtils.mergeStringArrays(null, taintArray())); // $hasTaintFlow
+
+    sink(StringUtils.parseLocale(taint()));
+
+    sink(StringUtils.parseLocaleString(taint()));
+
+    // sink(StringUtils.parseTimezoneString(taint())); // does not exist in spring 5.2.3
```
Comment (Marcono1234) [inline java/ql/test/library-tests/frameworks/spring/Test.java:74]: Are you sure that this does not exist? It is part of the stub, though the capitalization here is wrong (<code>Time<del>z</del><ins>Z</ins>one</code>):
```suggestion
    sink(StringUtils.parseTimeZoneString(taint())); // $hasTaintFlow
```
(Additionally its CSV entry is missing)

File: java/ql/src/semmle/code/java/frameworks/spring/SpringStringUtils.qll (hunk 7 above)
Comment (sauyon) [reply java/ql/src/semmle/code/java/frameworks/spring/SpringStringUtils.qll:38]: I just removed those two, I should probably remove `toLanguageTag` as well.

File: java/ql/test/library-tests/frameworks/spring/Test.java (hunk 8 above)
Comment (sauyon) [reply java/ql/test/library-tests/frameworks/spring/Test.java:74]: :facepalm: 
I removed this along with the two `parseLocale` functions.

**Comment by sauyon** [comment]
Superseded by https://github.com/github/codeql/pull/5976

**Pull request closed without merging**